RecoHI/HiJetAlgos/plugins/ParticleTowerProducer.cc

```diff
@@ -0,0 +1,216 @@
+// -*- C++ -*-
+//
+// Package:    ParticleTowerProducer
+// Class:      ParticleTowerProducer
+//
+/**\class ParticleTowerProducer ParticleTowerProducer.cc RecoHI/ParticleTowerProducer/src/ParticleTowerProducer.cc
+ Description: [one line class summary]
+ Implementation:
+     [Notes on implementation]
+*/
+//
+// Original Author:  Yetkin Yilmaz,32 4-A08,+41227673039,
+//         Created:  Thu Jan 20 19:53:58 CET 2011
+//
+//
+
+#include "RecoHI/HiJetAlgos/plugins/ParticleTowerProducer.h"
+
+#include "DataFormats/CaloTowers/interface/CaloTowerDefs.h"
+#include "DataFormats/CaloTowers/interface/CaloTowerDetId.h"
+#include "DataFormats/Candidate/interface/Particle.h"
+#include "DataFormats/Common/interface/Handle.h"
+#include "DataFormats/Common/interface/SortedCollection.h"
+#include "DataFormats/GeometryVector/interface/GlobalPoint.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/Framework/src/WorkerMaker.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescriptionFiller.h"
+#include "Geometry/Records/interface/CaloGeometryRecord.h"
+
+#include "TMath.h"
+
+#include <cmath>
+#include <cstdlib>
+#include <memory>
+#include <string>
+#include <vector>
+
+//
+// constants, enums and typedefs
+//
+
+//
+// static data member definitions
+//
+
+//
+// constructors and destructor
+//
+ParticleTowerProducer::ParticleTowerProducer(const edm::ParameterSet& iConfig) : geo_(nullptr) {
+  //register your products
+  src_ = consumes<reco::PFCandidateCollection>(iConfig.getParameter<edm::InputTag>("src"));
+  useHF_ = iConfig.getParameter<bool>("useHF");
+
+  produces<CaloTowerCollection>();
+
+  //now do what ever other initialization is needed
+}
+
+ParticleTowerProducer::~ParticleTowerProducer() {
+  // do anything here that needs to be done at desctruction time
+  // (e.g. close files, deallocate resources etc.)
+}
+
+//
+// member functions
+//
+
+// ------------ method called to produce the data  ------------
+void ParticleTowerProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  using namespace edm;
+
+  if (!geo_) {
+    edm::ESHandle<CaloGeometry> pG;
+    iSetup.get<CaloGeometryRecord>().get(pG);
+    geo_ = pG.product();
+  }
+
+  resetTowers(iEvent, iSetup);
+
+  edm::Handle<reco::PFCandidateCollection> inputsHandle;
+  iEvent.getByToken(src_, inputsHandle);
+
+  for (auto const& particle : *inputsHandle) {
+    double eta = particle.eta();
+
+    int ieta = eta2ieta(eta);
+    int iphi = phi2iphi(particle.phi(), ieta);
```

Comment:
Well, the original idea was to make a 2D grid in the physical eta-phi space, i.e., using the direction of the particles from the PV.   The idea of using a grid is basically to reuse the calorimetric style of background subtraction, approximating the same granularity to get the gross changes vs eta.  There is some smearing due to the difference between the physical eta-phi, and the values at the calorimeter face.  Perhaps there's a more precise way to do it,  but the subtraction seems to work rather well with this hybrid method.

Reply:
Thanks for clarifying. Some comment inline in the code would help.
On technical performance side (speed) it probably means that some simpler indexing (even a flat rectangular grid in eta-phi) may work. 
Remind me please where is this code expected to be running (to see if it can become CPU-limited)

Reply:
I believe the code is already running and has been for quite some time.  It's used in akPu4PFJets, which we are writing to AOD.

Reply:
and I thought that we do not have legacy modules running in reco anymore (ParticleTowerProducer is the legacy ```edm::EDProducer```).

Indeed, it's there, I found it in e.g. 140.55 in a test I have around from 11_1_0_pre7

Reply:
in that workflow I see PFTowers takes about 9 ms/ev; perhaps not something that will show up on a hotspot list even in light processing jobs

Reply:
> Indeed, it's there, I found it in e.g. 140.55 in a test I have around from 11_1_0_pre7

I'm slow to realize that this is in the HeavyIons scenario, while the later Run2 and Run3 plans are to run pp_on_AA setup.
Is there a plan to enable it there?

Reply:
akPu4PFJets was enabled for in the pp_on_AA setup for Run 2, so I believe this module was run.

Reply:
> akPu4PFJets was enabled for in the pp_on_AA setup for Run 2, so I believe this module was run.

I'm not sure if this is on the subject of ParticleTowerProducer.cc (a producer by itself)
```akPu4PFJets``` is made by FastjetJetProducer with MultipleAlgoIterator

Reply:
The input to Pu jets is PFTowers:
>>> process.akPu4PFJets.src
cms.InputTag("PFTowers")

>>> process.PFTowers
cms.EDProducer("ParticleTowerProducer",
    src = cms.InputTag("particleFlow"),
    useHF = cms.bool(False)
)

Reply:
> The input to Pu jets is PFTowers:

thank you.